### ISSUE TITLE
The example usage for macro is passing the type itself

### ISSUE THREAD

**Issue by jgabaut** [jgabaut/koliseo#49]
The `basic_example` provided is calling the `KLS_PUSH()` macro passing the `Example` type itself.

This is not recommended to so, since the call won't be correctly updated if you change the types and the `typedef` ends up having an incorrect size.

Using an expression pointer size is preferreable (eg. `*example` instead).

- `basic_example` should be fixed
- `README` should be updated
- `KLS_BASIC` backend calls should be fixed.

**Comment by jgabaut** [comment]
I honestly forgot what I meant and haven't been able to write down what is needed for this.

**Comment by khushal-banks** [comment]
1. `KLS_PUSH(kls, type)` defined in [koliseo.h +483](https://github.com/jgabaut/koliseo/blob/master/src/koliseo.h)
2.  and used in [basic_example.c +17](https://github.com/jgabaut/koliseo/blob/master/static/basic_example.c) as `KLS_PUSH(kls, Example, 1)` where Example is the type of an example struct.

- one KLS_PUSH expects 2 arguments and other one expects three arguments.
- Is basic_example.c is not updated? and should update to `KLS_PUSH(kls, Example)` instead?

> Using an expression pointer size is preferreable (eg. `*example` instead)

Do you mean?
```
// syntax of kls_push_zero_AR() declared in koliseo.h
extern void *kls_push_zero_AR(Koliseo *kls, size_t size, ptrdiff_t align, ptrdiff_t count);

// KLS_PUSH defined as following in koliseo.h
#define KLS_PUSH(kp, ep)              kls_push_zero_AR(kp, sizeof *ep, alignof *ep, 1)

// used as following in basic_example.c
Example* ep = (Example*) KLS_PUSH(kp, ep);
```

where,
- kp is pointer to buffer of type Koliseo
- and ep is pointer to buffer of type Example
- something i read [related article](https://pvs-studio.com/en/blog/posts/cpp/a0050/)

Is what i inferred is correct?

**Comment by jgabaut** [comment]
>     * Is basic_example.c is not updated? and should update to `KLS_PUSH(kls, Example)` instead?

Yes, `basic_example.c` was supposed to be updated with `0.4.0` to show the refactor that happened to all core macros.

> Is what i inferred correct?

About your reflection, I never questioned taking `ptrdiff_t` as size parameter since it came directly from the implementation of u/sketo and it seemed to me that, since the main push call is also expecting a "count" parameter, it's using `ptrdiff_t` to account for safely pushing big arrays where `PTRDIFF_MAX` may come into play (also mentioned in your link).

I honestly don't know if there would be any benefit to defining a wrapper layer that passed 1 to count and expects `size_t` as size. I don't know if it relates to this issue.

On this issue in particular, I don't remember why I never closed it. I definitely tried passing the same expression to the push call and it didn't work, I recall:

```c
[...]
Example* ep = KLS_PUSH(kls, ep);
```

This doesn't work.

I don't really know how to tackle the problem stated in top post, about changing the pointer size or something.

**Comment by khushal-banks** [comment]
> I honestly don't know if there would be any benefit to defining a wrapper layer that passed 1 to count and expects `size_t` as size. I don't know if it relates to this issue.

This wrapper already exist in code defined in [koliseo.h +483](https://github.com/jgabaut/koliseo/blob/master/src/koliseo.h). I didn't redefine it.


> On this issue in particular, I don't remember why I never closed it. I definitely tried passing the same expression to the push call and it didn't work

I am de referencing the pointer when using sizeof and alignof which will give expected output. Maybe you missed to de-reference it (if missed sizeof and alignof would give values based on size of a pointer which is fix)

**Comment by khushal-banks** [comment]
> I don't really know how to tackle the problem stated in top post, about changing the pointer size or something.

Can i try it as well? I am guessing you would want to see these changes in devel branch, right?
Is this project open for contributions?

**Comment by khushal-banks** [comment]
> A properly defined function like `void* my_func(Koliseo* kls, size_t size)` that internally goes to `kls_push_zero_AR()`

Wao :+1: I agree. Most of macros in koliseo.h should also be avoided in the similar manner, right?

> please. I'd like contributions

I was mostly messing with you and myself. I would try to do something if i felt up to the challenge. But this feels so complicated to me. Mostly because all functions are named similar to each other and there are so many of these functions that i don't understand what to use.

Plus, i have no background info in animation like what is an arena allocator. :) 
Every time i read this file, arena seems like a boxing ring and koliseo the man who is beating me to death.

> or you want to refactor the core calls to take `size_t`

Is this not right? I mean the value we pass in size is returned by sizeof.

sizeof returns size_t. so value we pass is size_t in real. ptrdiff_t is what we are type-casting it to (which i believe is bigger in size) but overflowed values from size_t will not correctly reflect in ptrdiff_t. This would definitely lead to trouble.

I am just writing it to understand by discussion. I personally never used ptrdiff_t. But i strongly believe i am on the right track here. But i always believe that so i might be wrong.

**Comment by jgabaut** [comment]
> Most of macros in koliseo.h should also be avoided in the similar manner, right?

I guess it's not strictly needed. Macros let the user code not worry about what' the actual internal code, **if** we commit to not break the macros (which was done on `0.4.0`). I considered offering signatures without the count, but didn't really want to mantain another function, for however thin it might be (even just straight calling the underlying function).


> Mostly because all functions are named similar to each other and there are so many of these functions that i don't understand what to use.

Almost all functions start with `kls_` to namespace them correctly in user code. Stealing `new()` for the library is a major symbol theft offence. Straight to jail. (/s).

If that's not what you meant by them being so similar, I guess you're referring to the myriad of various `push()` calls.

They get increasingly custom. That's why the average user will stop at the functions shown in `basic_example.c`, without even trying to configure the arena.

> Plus, i have no background info in animation like what is an arena allocator. :)

This repo doesn't have anything to do with animation. I'll lend you an interesting link that I read while starting this: [link](https://www.gingerbill.org/article/2019/02/08/memory-allocation-strategies-002/). This explains the idea of an arena better than how I could here going for a long post.

> :) Every time i read this file, arena seems like a boxing ring and koliseo the man who is beating me to death.

The name "koliseo" is the esperanto rendition of the Colosseum, which you may not be familiar with: [link](https://en.wikipedia.org/wiki/Colosseum).

> sizeof returns size_t. so value we pass is size_t in real. ptrdiff_t is what we are type-casting it to (which i believe is bigger in size) but overflowed values from size_t will not correctly reflect in ptrdiff_t. This would definitely lead to trouble.

I guess you might be right.

The only thing that matters is passing `*ep` instead of `Example`, and it can be done easily at least for the first two points of top post.

If you want to contribute, even a single-line fix for the bug you found in `basic_example.c` would be enough (referring to the separate issue #65). Apart from that, you could also update the calls to not use the type itself but the deref-pointer expression, as said before.

I'll get around to doing these myself soon, I'll wait to see if you want to do it and get included in next release.

**Comment by khushal-banks** [comment]
> The `basic_example` provided is calling the `KLS_PUSH()` macro passing the `Example` type itself.
> 
> Using an expression pointer size is preferreable (eg. `*example` instead).

Hello @jgabaut, i tried this.
Problem is `_Alignof` or alternative `alignof` both takes `type` as first argument.
`KLS_PUSH` macro and `kls_push_zero_AR` function **requires** evaluated `align` value *for padding and offset calculations.*

Here is the list of some more similar functions which requires `align` value for same calculation.
- kls_push_zero_named
- kls_push_zero_typed
- kls_temp_push_zero_named
- kls_temp_push_zero_typed
- kls_pop_AR
- kls_temp_pop_AR

**Comment by khushal-banks** [comment]
type is also used by `STRIGIFY` macro to supply `desc` argument to `kls_push_zero_typed` and `kls_push_zero_named` functions.

Although known types can be found using `_Generic` [reference](https://stackoverflow.com/questions/9804371/syntax-and-sample-usage-of-generic-in-c11/17290414#17290414)

But, this wouldn't be a our solution because we need to support user-defined custom types as well

**Comment by khushal-banks** [comment]
C23 may remove `alignof` operator. But `_Alignof` would stay, if i understood right. Check the link below to be sure.
Type support recent changes/plans: [types](https://en.cppreference.com/w/c/types)

> Does it depend on the basic type behind the allocation?

[Yes](https://en.cppreference.com/w/c/language/object#Alignment)
align is basically the size of biggest basic data-type in structure if i understood it right.

> Compiling this lib with `--std=c99`, is it doable?

Not sure. I will have to improve my understanding of this code. But maybe one could calculate padding/offset using something like [this](https://en.cppreference.com/w/c/types/offsetof)

> The "internal" pointer is "*char", which can alias to anything

Do you mean internal type is void*? This should be okay. It is typical to return void*

> push macros are all doing an implicit cast of the returned pointer

macros can type-cast but it doesn't look pretty (my personal take). But they can. It should be legal C. But, i am sure you know most of these macros shouldn't even exist. They are basically wrappers to 2-4 functions and the little changes i did while trying to remove `type` requirement; i felt like all/most of these macros can be removed. But again not really sure.
I agree that it would be best if caller did the type-casting rather than macros.

# More details i observed:
1. I previously mentioned that sizeof would return size_t which cann't be type-casted to ptrdiff_t. New observation is that they differ in sign as well. I think size_t is unsigned long and ptrdiff_t is signed long. So, typecasting is more illegal.
2. One more new observation is that `_Alignof` returns size_t. I confirmed it from [here](https://en.cppreference.com/w/c/language/_Alignof) So, there should be no ptrdiff_t in this code as far as i understand. Please confirm this.

**Comment by jgabaut** [comment]
> Yes align is basically the size of biggest basic data-type in structure if i understood it right. [Check](https://en.cppreference.com/w/c/language/object#Alignment)

Looking at [this](https://stackoverflow.com/questions/38271072/how-to-determine-maximum-required-alignment-in-c99), it seems there's some way to get the maximum supported alignment in `C99`. Handling this would need some tests to be added, for the edge cases (?). Although, if I am not mistaken, these would mostly matter on esoteric platforms, right?

> > Compiling this lib with `--std=c99`, is it doable?
> 
> Not sure. I will have to improve my understanding of this code. But maybe one could calculate padding/offset using something like [this](https://en.cppreference.com/w/c/types/offsetof)

This would not help in building with `--std=c99`, since it's `C23`.

> 
> > The "internal" pointer is "*char", which can alias to anything
> 
> Do you mean internal type is void*? This should be okay. It is typical to return void*

No, I mean the backing arena memory buffer is `char*` and the returned pointer is into that region.
`*char` can alias to anything, but I can't find links about this. I think it's exempt from alignment issues, or something.

> They are basically wrappers to 2-4 functions and the little changes i did while trying to remove `type` requirement; 

Their purpose is providing an easy call into the deeper function. If you want to propose a different upper interface for the `KLS_PUSH()` macros, propose one, I'd like to discuss. The main trick they do is obviously the implicit `sizeof`.

>     1. I previously mentioned that sizeof would return size_t which cann't be type-casted to ptrdiff_t. New observation is that they differ in sign as well. I think size_t is unsigned long and ptrdiff_t is signed long. So, typecasting is more illegal.

The different in signedness is not necessarily a bad thing. The important question is: what is the pathological input that would cause `sizeof` result (unsigned) to be not representable in `ptrdiff_t` (so, if I got this right, bigger than `PTRDIFF_MAX`)?

If the passed size is too big to be handled by `ptrdiff_t`, it's out of the scope of the library. But I'm still wondering if the too-large-size would be caught inside of the macro?

Also, does the distinction between `_ARR` macros (expecting a count) and plain calls (implicit count of 1) matter in this case? I think it may.

>     2. One more new observation is that `_Alignof` returns size_t. I confirmed it from [here](https://en.cppreference.com/w/c/language/_Alignof) So, there should be no ptrdiff_t in this code as far as i understand. Please confirm this.

`ptrdiff_t` is something I was unfamiliar with before reading some u/skeeto code, and even though I may not have the strongest arguments here to support this, I am sure it is the type used for a specific reason.

I suspect the key lies around `PTRDIFF_MAX` and the mentioned pathological construct, any ideas?

**Comment by khushal-banks** [comment]
I agree with everything you said in last comment. except that `offsetof` is not C23. I think it is available in all standards. Custom definition is also possible [link](https://en.wikipedia.org/wiki/Offsetof) and [max_align_t](https://en.cppreference.com/w/c/types/max_align_t) looks like a const value of alignment 8 or 16 (based on architecture), corresponding to what is alignment of intmax_t which is defined as long long or long double. so that shouldn't be useful for our requirement.

> The important question is: what is the pathological input that would cause `sizeof` result (unsigned) to be not representable in `ptrdiff_t`

I would recommend adding a test with **struct size** greater than `size_t` and less than `ptrdiff_t` (if it doesn't exist right now). That should confirm that there are no issues when using ptrdiff_t

**Comment by jgabaut** [comment]
> except that `offsetof` is not C23. I think it is available in all standards. 

I think you may be closer to the truth than me. I am now suspecting it was standardised like it it now in `C99`.

Still, I don't think knowing the field offset would help to know the whole alignment. 
Maybe something more to say about asking for the "0" field's offset size, to know the whole size? I may be misunderstanding this.

 > max_align_t looks like a const value of alignment 8 or 16 (based on architecture),

That may be right, but `max_align_t` is `C11`.

> so that shouldn't be useful for our requirement.

If one wanted to drop the `_Alignof` usage and build with `C99`, maybe the passed alignment would be substituted for the largest possible one (to be safe, we use the biggest alignment). 

On this endeavour, my previous reply hosts a couple comments saying that the best way to handle this in `C99` is to compare the largest possible sizes (they provide a list) and use the biggest.

> I would recommend adding a test with **struct size** greater than `size_t` and less than `ptrdiff_t` (if it doesn't exist right now). That should confirm that there are no issues when using ptrdiff_t
> 

How big would one such structure be?

Also, would a `size_t` bigger than `PTRDIFF_MAX` result in a negative `ptrdiff_t` value?

I don't remember the checks done on the requested size in the functions, I guess  negative ones and ones over `PTRDIFF_MAX` should be rejected already. I'll have to check the code, you may also do so if you want.

**Comment by khushal-banks** [comment]
> I suspect the key lies around `PTRDIFF_MAX`
> If the passed size is too big to be handled by `ptrdiff_t`, it's out of the scope of the library.

Okay, now i understand .. there is no issue in using `ptrdiff_t` if we are not considering values bigger than PTRDIFF_MAX. This should work perfectly.

> Still, I don't think knowing the field offset would help to know the whole alignment.

Yeah. I agree. So, current implementation of KLS_PUSH taking `type` as argument should be fine. Supporting C11 and above should be fine as well.

**Comment by jgabaut** [comment]
> Okay, now i understand .. there is no issue in using `ptrdiff_t` if we are not considering values bigger than PTRDIFF_MAX. As, size_t and ptrdiff_t have same number of bytes. Difference is only in unsigned/signed ness. This should work perfectly.

I'd like to still include a test case for a "bad" allocation, if doable.
There's already some tests present, you should be able to run them with `make anviltest`.

> So, current implementation of KLS_PUSH taking `type` as argument should be fine.

Maybe, but the problem stated in top post is still not caught by the compiler, if I'm not mistaken.

Still, two ideas came out of this:
- Consider dropping the implicit cast in `_PUSH` macros
- Try providing a macro not using `_Alignof()` in order to build with `C99`
  - Perhaps the best way might be always passing [the largest possible alignment requirement for current platform](https://stackoverflow.com/questions/38271072/how-to-determine-maximum-required-alignment-in-c99)

I'll open other issues and see what can be done.

Returning to the matter of this issue, I haven't confirmed that `_Alignof(*example)` does not work, but I trust you tried... Is there any solution, or is this a forced evil we must declare?

**Comment by khushal-banks** [comment]
> Maybe, but the problem stated in top post is still not caught by the compiler, if I'm not mistaken
> **TOP-POST:**
> since the call won't be correctly updated if you change the type of the declaration and the `typedef` ends up having an incorrect size

If user changes the type of the declaration then user would need to recompile.
`typedef` would be updated to correct size. Did i miss something?

**Comment by jgabaut** [comment]
> `typedef` would be updated to correct size. Did i miss something?

I was thinking of this:

```c
Foo* ex = KLS_PUSH(kls, Foo);
```

Being erroneously changed to look like this (dev forgets to update right side):
```c
Bar* ex = KLS_PUSH(kls, Foo);
```

When `sizeof(Foo) != sizeof(Bar)`

**Comment by jgabaut** [comment]
@khushal-banks

I noticed in the failed workflow that the build error for `_Alignof(expression)` comes from `pedantic` flag.

Does this mean there's some extensions (not-ISO C) out there allowing this idiom?
